itsValyria/Oncollaboration
# Search #25

## What does this change?

### Search 
 
This feature branch was created to make a search results component, and to make the search function actually work. Previously it was just plain html and css, it did not function. This function is necessary because users can search for certain webinars or contourings.
 
I am using a function named ```handleSearch``` to redirect from any page that uses the search bar, to the search page, this function is called when the user submits the search form.

```js
  import { goto } from '$app/navigation';
  let searchQuery = '';

  // Redirect to the /search page with the URL-encoded search query
  function handleSearch() {
  goto(`/search?query=${encodeURIComponent(searchQuery)}`);
}
```

To fetch all the data for the contourings and webinars from Directus, and filter those contourings and webinars, I have a load function that does the following:

1. Extracting the query parameter from the URL and converting it to lowercase.
2. Fetching webinar and contouring data from their API endpoints.
3. Filtering the data to only include items whose titles match the search query
4. Returning the filtered results along with the original query for use in the front-end

```js
export async function load({ url }) {
  // Converts the query string retrieved from the URL to lowercase
  const query = url.searchParams.get('query')?.toLowerCase() || '';
  
  // Define API endpoint URLs to fetch webinars en countourings data
  const webinar = `${baseURL}avl_webinars?fields=*.*.*&sort[]=-date`;
  const contouring = `${baseURL}avl_contourings?fields=*.*.*`;
  
  // Fetch the data from their API endpoints
  const webinars = await fetchJson(webinar);
  const contourings = await fetchJson(contouring);

  // Filter webinars based on the query input
  const filteredWebinars = webinars.data.filter(webinar =>
    webinar.title.toLowerCase().includes(query)
  );

  // Filter contourings based on the query input
  const filteredContourings = contourings.data.filter(contouring =>
    contouring.title.toLowerCase().includes(query)
  );

  // Returning the query back to the front-end along with the filtered webinars and contourings
  return {
    webinars: filteredWebinars,
    contourings: filteredContourings,
    query
  };
}
```

When you search, you end up on a seperate page with route ```/search?query={inputHere}```. I chose to do this because you can have quite a lot of results when there are hundreds of webinars or contourings. This page shows the results of the search query input. If one turns out to be empty, the section completely disappears. I'm doing this using if statements.

```html
<main>
  <h1>Search Results</h1>
  <Search></Search>
  <p>Results for: <strong>"{data.query}"</strong></p>
  
  {#if data.webinars.length > 0}
    <h2>Matching Webinars</h2>
    <div class="carrousel">
      {#each data.webinars as webinar}
        <div>
          <WebinarOverview {...webinar} />
        </div>
      {/each}
    </div>
  {/if}

  {#if data.contourings.length > 0}
    <h2>Matching Contourings</h2>
    <div class="carrousel">
      {#each data.contourings as contouring}
        <div>
          <ContouringOverview {...contouring} />
        </div>
      {/each}
    </div>
  {/if}

  {#if data.webinars.length === 0 && data.contourings.length === 0}
    <p>No results found.</p>
  {/if}
</main>
```



 
 
## How Has This Been Tested?
 

### Accessibility

The first accessibility test I did gave a score of **94**, because the search button did not have an accessible name. I fixed this by changing this line of code:

```html
// Before
<button type="submit">
// After
<button type="submit" name="search" id="search" aria-label="Search">
```

After doing this correction the lighthouse test gave a score of **100**.

### Performance

The performance has a score of 55, which is not that much. We can fix this by using directus custom transformation on the images to convert them to ```.webp```, I made a seperarte issue for that, since it isn't an issue caused by this specific component. (#40)

### Responsive

The website is fully responsive. No bugs were found, except for REALLY small screens (<200px width)

### Device

I tested on my iPad, my iPhone, my MacBook and my PC. I did not found any bugs or any other things I have to fix.

### Browser

**1. Chrome**
Chrome is my default browser, I use this to develop. Therefore, I found no existing bugs in the chrome browser.

**2. Firefox**
Firefox works like a charm, just like chrome. Everything does what it should do.

**3. Safari**
Safari also works!! No bugs found.

**4. Edge**
Edge also works!! No bugs found hehe.
 
## UI Changes
 
I decided to align the search bar on the left, along with the rest of the website. This gives a more uniform and clean look.

### Before:

![image](https://github.com/user-attachments/assets/1509d412-77e8-4162-8347-35c1014dc799)

### After:

![image](https://github.com/user-attachments/assets/c6e56ec3-e72f-4bd1-b984-81026e8917c2)
 
## How to review
 
 1. Download code and run it locally
 2. Use the search bar, try a few different search terms
 3. Witness if the search actually works
 4. Witness if you only see matching webinars when there are no matching contourings, and vice versa.
 5. Witness if you can correctly see the term you searched for
